### DISCUSSION TITLE
How to update a many to many relationship?

### DISCUSSION THREAD

**Discussion by Bradleykingz** [prisma/prisma#3283] [Q&A] [answered]
I'm having a hard time figuring out why my many-to-many relationship doesn't update. 

```prisma
model post {
  id       Int        @default(autoincrement()) @id
  content  String?
  categories category[] @relation("post_categories")
}

model category {
  id   Int     @default(autoincrement()) @id
  name String? @unique
  posts post[]  @relation("post_categories")
}
```

With the above Prisma schema, I ran the following code:

```js
import {PrismaClient} from './prisma/generated/client'

const prisma = new PrismaClient()

// create a post
async function create() {
    return prisma.post.create({
        data: {
            content: "lots of content",
            categories: { // create one category
                create: {
                    name: "Big fat cats"
                }
            }
        },
    });
}

async function update(id: number) {

    // create category
    const category1 = await prisma.category.create({
        data: {
            name: "Small pretty dandelions"
        }
    })
   
   // create a second category
    const category2 = await prisma.category.create({
        data: {
            name: "Yellow sunflowers"
        }
    })

    return prisma.post.update({
        where: {
            id
        },
        data: {
            categories: {
                connect: [category1, category2].map(cat=> ({id: cat.id}))
            }
        },
        include: {
            categories: true
        }
    })
}

async function main() {
    const post = await create();
    return update(post.id);
}

main().then(res=> {
    console.log("Updated post", res);
}).catch(err=> {
    console.error(err);
})
```

But the above log results in 

``` 
Updated post {
  id: 1,
  content: 'lots of content',
  categories: [ { id: 1, name: 'Big fat cats' } ]
}
```

All three categories are created fine. However, the last two are not 'connected' to the **post** object as I would expect to happen. Am I doing something wrong?

**Comment by ryands17** [comment, marked as answer]
Hey @Bradleykingz 👋 
Remove `async` from the `create` function and run it again and check.

**Reply by Bradleykingz** [reply]
Nope. No luck.

**Reply by ryands17** [reply]
This works for me:
```ts
function create() {
  return prisma.post.create({
    data: {
      content: 'lots of content',
      categories: {
        // create one category
        create: {
          name: 'Big fat cats',
        },
      },
    },
  })
}

async function update(id: number) {
  // create category
  const category1 = await prisma.category.create({
    data: {
      name: 'Small pretty dandelions',
    },
  })

  // create a second category
  const category2 = await prisma.category.create({
    data: {
      name: 'Yellow sunflowers',
    },
  })

  return prisma.post.update({
    where: {
      id,
    },
    data: {
      categories: {
        connect: [category1, category2].map((cat) => ({ id: cat.id })),
      },
    },
    include: {
      categories: true,
    },
  })
}

async function main() {
  const post = await create()
  return update(post.id)
}

main()
  .then((res) => console.log(JSON.stringify(res, null, 2)))
  .catch(console.error)
  .finally(async () => {
    await prisma.$disconnect()
  })
```

![image](https://user-images.githubusercontent.com/19697099/90132527-a5cba200-dd8b-11ea-84f3-66c68e6a9d95.png)

**Reply by Bradleykingz** [reply]
Weird. Here's what it returns on my machine.

![Screenshot from 2020-08-13 15-10-11](https://user-images.githubusercontent.com/19578478/90132848-2d5ae600-dd77-11ea-8500-25b12ea5d6f1.png)

**Reply by Bradleykingz** [reply]
Is there any particular reason why it should work without async?

**Reply by ryands17** [reply]
You're not using `await` in the function so `async` ins't needed there actually. Also I'm on version *2.4.1* of `@prisma/client` and `@prisma/cli` if that helps.

**Reply by Bradleykingz** [reply]
I'm also using the latest versions but still no luck.

When I change it to 

```js
function create() {
    return prisma.post.create({
        data: {
            content: 'lots of content',
        },
    })
}
```
Only one item is updated

```json
{
  "id": 1,
  "content": "lots of content",
  "categories": [
    {
      "id": 1,
      "name": "Small pretty dandelions"
    }
  ]
}

```
Is there an alternative way I could approach this?

**Reply by ryands17** [reply]
Did you check the database if the relations are updated?
If not, create an issue [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=bug_report.md&title=) as I am not able to reproduce this.

**Reply by Bradleykingz** [reply]
Actually, I've found the issue.

Here's how I was creating my relation table

```sql
create table _post_categories
(
    "A" integer unique not null references category("id"),
    "B" integer unique not null references post ("id")
);

create unique index "idx_category_posts_unique" on _post_categories ("A", "B");
create index "idx_category_posts" on _post_categories ("B");
```

Removing 'unique' from both A & B sorted it out.

Thanks for the help!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
